File: src/main/scala/scaled/pacman/Depend.scala
//
// Scaled - a scalable editor extensible via JVM languages
// http://github.com/samskivert/scaled/blob/master/LICENSE

package scaled.pacman

// TODO: other scopes?
sealed trait Scope
case object Compile extends Scope
case object Test extends Scope

// TODO: model kind?
case class RepoId (
  groupId :String, artifactId :String, version :String, kind :String, scope :Scope
) extends Package.Id

sealed trait Depend
case class SourceDepend (source :Source) extends Depend
case class MavenDepend (id :RepoId) extends Depend
case class IvyDepend (id :RepoId) extends Depend

/** [[Depend]] related utilities. */
object Depend {

  /** Parses a string representation of a [[Depend]]. */
  def parse (url :String) :Depend = url.split(":", 2) match {
    case Array("mvn", rest) => MavenDepend(parseRepoId(rest))
    case Array("ivy", rest) => IvyDepend(parseRepoId(rest))
    case Array(vcs,   url ) => SourceDepend(Source.parse(vcs, url))
  }

  // parses a repo depend: repo:groupId:artifactId:version:kind:scope
  private def parseRepoId (text :String) :RepoId = text.split(":") match {
    case Array(groupId, artifactId, version, kind, scope) =>
      RepoId(groupId, artifactId, version, kind, parseScope(scope))
    case Array(groupId, artifactId, version, kind) =>
      RepoId(groupId, artifactId, version, kind, Compile)
    case Array(groupId, artifactId, version) =>
      RepoId(groupId, artifactId, version, "jar", Compile)
    case other => throw new IllegalArgumentException(
      s"Invalid repo id: $text (expect 'groupId:artifactId:version:scope')")
  }

  private def parseScope (scope :String) = scope match {
    case "compile" => Compile
    case "test" => Test
    case _ => throw new IllegalArgumentException(s"Invalid scope: $scope")
  }
}

File: src/main/java/scaled/pacman/Depend.java
//
// Scaled - a scalable editor extensible via JVM languages
// http://github.com/samskivert/scaled/blob/master/LICENSE

package scaled.pacman;

import java.net.URISyntaxException;

public class Depend {

  public static interface Id {
  }

  // TODO: other scopes?
  public static enum Scope {
    COMPILE, TEST;
  }

  /** Parses a string representation of a [[Depend]]. */
  public static Depend parse (String url, Scope scope) throws URISyntaxException {
    String[] bits = url.split(":", 2);
    if (bits.length == 1) throw new IllegalArgumentException("Invalid depend URI: " + url);
    return new Depend(parseId(bits[0], bits[1]), scope);
  }

  private static Id parseId (String tag, String data) throws URISyntaxException {
    switch (tag) {
      case "mvn": return RepoId.parse(data);
      case "sys": return SystemId.parse(data);
      default:    return Source.parse(tag, data);
    }
  }

  public final Id id;
  public final Scope scope;

  public Depend (Id id, Scope scope) {
    this.id = id;
    this.scope = scope;
  }

  /** Returns true if this is a source depend. */
  public boolean isSource () {
    return (id instanceof Source);
  }

  @Override public String toString () { return id + ":" + scope.toString().toLowerCase(); }
  @Override public int hashCode () { return id.hashCode() ^ scope.hashCode(); }
  @Override public boolean equals (Object oo) {
    return (oo instanceof Depend) && ((Depend)oo).id.equals(id) && ((Depend)oo).scope == scope;
  }
}
